### PR TITLE
Throttle dialogflow calls.

### PR DESCRIPTION
**Pull request by SurferJeffAtGoogle**
Also, RetryRobot makes dialogflow tests cope with eventual consistency in the list() method.

### CODE REVIEW

**Comment by beccca** [comment]
TIL about `BlockingCollection` and it's pretty darned cool!

Took me a couple mins to walk through and grok it. I can't think of any semantics which would've helped me grok it faster (except 1 comment I made about `timeSpan`, feel free to make this change but I'm LGTM!)

File: dialogflow/api/Test/DialogflowTest.cs [hunk 1]
```diff
@@ -72,4 +84,50 @@ public ConsoleOutput RunWithSessionId(string command, params object[] args)
...
+    // TODO: Move this class into test helpers.
+    /// <summary>
+    /// Schedules throttling.
+    /// </summary>
+    class ThrottleTokenPool
+    {
+        private readonly TimeSpan _timeSpan;
+        private readonly BlockingCollection<ThrottleToken> _pool =
+            new BlockingCollection<ThrottleToken>();
+        public ThrottleTokenPool(int tokenCount, TimeSpan timeSpan)
```
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply dialogflow/api/Test/DialogflowTest.cs]: Added lots of comments.  Thank you for the review!

File: dialogflow/api/Test/DialogflowTest.cs [hunk 2]
```diff
@@ -72,4 +84,65 @@ public ConsoleOutput RunWithSessionId(string command, params object[] args)
             return Run(command, arguments.ToArray());
         }
     }
+
+    // TODO: Move this class into test helpers.
+    /// <summary>
+    /// Schedules throttling.
+    /// </summary>
+    class ThrottleTokenPool
```
Comment (simonz130) [inline dialogflow/api/Test/DialogflowTest.cs]: Should the pool be static? Can you have multiple instances of throttler pools that in fact block requests to the same actual resource?

Comment (simonz130) [reply dialogflow/api/Test/DialogflowTest.cs]: I see that you create a static instance of the pool earlier, but IMO the class itself should also be static and should expose methods to initialize and get access to the throttle pool per resource type that is being throttled.

**Pull request merged**